src/gestures/use-pan-gesture.ts

```diff
@@ -397,150 +394,118 @@ export function usePanGesture(
 
-    const [startPointerUp, stopPointerUp] = usePointerEvents(
-        { onPointerUp },
-        safeWindow
-    )
+        // Only start panning if the offset is larger than 3 pixels. If we make it
+        // any larger than this we'll want to reset the pointer history
+        // on the first update to avoid visual snapping to the cursoe.
+        const distancePastThreshold = distance(info.offset, { x: 0, y: 0 }) >= 3
 
-    const [startPointerMove, stopPointerMove] = usePointerEvents(
-        { onPointerMove },
-        safeWindow,
-        { capture: true }
-    )
+        if (!panStarted && !distancePastThreshold) return
 
-    const onPointerDown = useCallback(
-        (event: MouseEvent | TouchEvent | PointerEvent, info: EventInfo) => {
-            // If we have more than one touch, we don't want to start detecting this gesture.
-            if (isTouchEvent(event) && event.touches.length > 1) {
-                return
-            }
-
-            const initialInfo = transformPoint(info)
-            const { point } = initialInfo
-
-            pointer.current = {
-                x: motionValue(point.x),
-                y: motionValue(point.y),
-            }
-
-            const { timestamp } = getFrameData()
-            session.current = {
-                target: event.target,
-                pointHistory: [{ ...point, timestamp }],
-            }
-
-            if (onPanSessionStart) {
-                onPanSessionStart(event, getPanInfo(initialInfo))
-            }
-
-            startPointerMove()
-            startPointerUp()
-        },
-        [onPointerUp, onPointerMove]
-    )
+        const { point } = info
+        const { timestamp } = getFrameData()
+        session.current.pointHistory.push({ ...point, timestamp })
+
+        if (!panStarted) {
+            onPanStart && onPanStart(lastMoveEvent.current, info)
+            session.current.startEvent = lastMoveEvent.current
+        }
+
+        onPan && onPan(lastMoveEvent.current, info)
+    }
+
+    function onPointerMove(
+        event: MouseEvent | TouchEvent | PointerEvent,
+        info: EventInfo
+    ) {
+        lastMoveEvent.current = event
+        lastMoveEventInfo.current = transformPoint(info)
+
+        // Because Safari doesn't trigger mouseup events when it's above a `<select>`
+        if (isMouseEvent(event) && event.buttons === 0) {
+            onPointerUp(event, info)
+            return
+        }
+
+        // Throttle mouse move event to once per frame
+        sync.update(updatePoint, true)
+    }
+
+    function onPointerUp(
+        event: MouseEvent | TouchEvent | PointerEvent,
+        info: EventInfo
+    ) {
+        cancelPan()
+
+        if (!session.current) {
+            warning(false, "onPointerUp fired without pointer session")
+            return
+        }
+
+        onPanEnd && onPanEnd(event, getPanInfo(transformPoint(info)))
+
+        session.current = null
+    }
+
+    function onPointerDown(
```

Comment:
I no this isn't much different from what we did before, but isn't creating all these inline functions wasteful?

Reply:
We were already creating them:

```jsx
useCallback(() => {}, [...deps])
```

Is still creating the function. The only performance benefit is not letting that lead to a re-render. But you also have to compare the deps, so it could actually lead to worse performance.

Now there's an early return this shouldn't be a problem but if it is we can look into other ways of loading things like this